network/src/dms.rs

```diff
@@ -1,65 +1,151 @@
+
+#[derive(Debug, Clone, Serialize)]
+pub struct Message {
+    data: String,
+    signature: TypedSignature<String>,
+}
+
+impl Message {
+    pub fn new(data: String, signature: TypedSignature<String>) -> Result<Self, CryptoError> {
+        signature.verify(&data)?;
+        Ok(Self { data, signature })
+    }
+
+    pub fn data(&self) -> &str {
+        &self.data
+    }
+
+    pub fn signature(&self) -> &TypedSignature<String> {
+        &self.signature
+    }
+}
+
+impl ToHash256 for Message {
+    fn to_hash256(&self) -> Hash256 {
+        Hash256::hash(serde_json::to_vec(self).unwrap())
+    }
+}
+
+/// A message before verification.
+#[derive(Debug, Clone, Serialize, Deserialize)]
+struct RawMessage {
+    pub data: String,
+    pub signature: TypedSignature<String>,
+}
+
+impl RawMessage {
+    #[allow(dead_code)]
+    pub fn into_message(self) -> anyhow::Result<Message> {
+        Message::new(self.data, self.signature).map_err(|e| anyhow!(e))
+    }
+
+    #[allow(dead_code)]
+    pub fn from_message(message: Message) -> Self {
+        RawMessage {
+            data: message.data().to_owned(),
+            signature: message.signature().clone(),
+        }
+    }
+}
 
 /// The interface that will be wrapped into an HTTP RPC server for the peers.
-#[async_trait]
-pub trait DistributedMessageSetRpcInterface {
+#[serde_tc_full]
+trait DistributedMessageSetRpcInterface: Send + Sync + 'static {
     /// Returns the messages except `knowns`. If the height is different, it returns `Err(height)`.
     async fn get_message(
         &self,
         height: BlockHeight,
         knowns: Vec<Hash256>,
-    ) -> Result<Vec<Message>, BlockHeight>;
+    ) -> Result<Vec<RawMessage>, String>;
 }
 
-pub struct DistributedMessageSetImpl {}
+/// A **cumulative** set that is shared in the p2p network, backed by the local file system.
+///
+/// One of the notable characteristics of blockchain is that it is based on heights;
+/// The key idea here is that we retain an instance (both in memory or on disk)
+/// of `DistributedMessageSet` only for a specific height,
+/// and discard if the height progresses, creating a new and empty one again.
+///
+/// For every methods,
+/// - If the given directory is empty, it fails (except `create()`).
+/// - It locks the storage.
+/// - If the given directory is locked (possibly by another instance of `DistributedMessageSet`),
+/// it will `await` until the lock is released.
+pub struct DistributedMessageSet<N, S> {
+    _storage: Arc<RwLock<S>>,
+    _marker: std::marker::PhantomData<N>,
+}
 
-#[async_trait]
-impl DistributedMessageSet for DistributedMessageSetImpl {
-    async fn create(_storage_directory: &str, _height: u64) -> Result<(), Error> {
+impl<N: P2PNetwork, S: Storage> DistributedMessageSet<N, S> {
+    /// Creates a new and empty storage with the given directory.
+    /// If there is already a directory, it discards everything and creates a new one.
+    /// You should try `open()` first!
```

Comment:
Anyway to improve this in a way s.t. opening an existing directory as default?

Reply:
that's not the role of this interface, but the role of the user.